### ISSUE TITLE
transparent proxy causing gnutls_handshake() fail on server

### ISSUE THREAD

**Issue by RadarNyan** [yrutschle/sslh#109]
I have setup sslh with transparent proxy, everything seems to work as expected: I can access my site on port 443 with ssl enabled, and nginx can get the real IP address. Until today I found that if transparent proxy is enabled, I'll encount problem if I try to access my site through a proxy on my server.

Heres what happens when I **don't use transparent proxy**:
```bash
root@ubuntu:~# sslh-select -u nobody --listen <my_public_ip>:443 --anyprot 172.16.0.1:443

root@ubuntu:~# curl -v https://<my_public_ip>/
*   Trying <my_public_ip>...
* Connected to <my_public_ip> (<my_public_ip>) port 443 (#0)
* found 173 certificates in /etc/ssl/certs/ca-certificates.crt
* found 692 certificates in /etc/ssl/certs
* ALPN, offering http/1.1
* SSL connection using TLS1.2 / ECDHE_RSA_AES_256_GCM_SHA384
*        server certificate verification OK
*        server certificate status verification SKIPPED
* SSL: certificate subject name (<my_domain_name>) does not match target host name '<my_public_ip>'
* Closing connection 0
curl: (51) SSL: certificate subject name (<my_domain_name>) does not match target host name '<my_public_ip>'
```
nginx listens 172.16.0.1:443 , as you can see SSL works fine (well, except my certificate is for my domain name not for my ip, that's not important here)

But if I **enable transparent proxy**:
```bash
root@ubuntu:~# sslh-select -u nobody --listen <my_public_ip>:443 --anyprot 172.16.0.1:443 --transparent

root@ubuntu:~# curl -v https://<my_public_ip>/
*   Trying <my_public_ip>...
* Connected to <my_public_ip> (<my_public_ip>) port 443 (#0)
* found 173 certificates in /etc/ssl/certs/ca-certificates.crt
* found 692 certificates in /etc/ssl/certs
* ALPN, offering http/1.1
* gnutls_handshake() failed: The TLS connection was non-properly terminated.
* Closing connection 0
curl: (35) gnutls_handshake() failed: The TLS connection was non-properly terminated.
```

If I try to run ```curl -v https://<my_public_ip>/``` from anywhere else but my server, for example from my home wifi, SSL works just fine. It just doesn't work if I try it on the server.

I have totally no idea what's going on, how can I fix this?
BTW this is the guide I followed to setup transparent proxy: http://rutschle.net/pipermail/sslh/2014-September/000547.html

**Comment by yrutschle** [comment]
On Mon, Nov 07, 2016 at 06:50:42AM -0800, RadarNyan wrote:
> If I try to run ```curl -v https://<my_public_ip>/``` from anywhere else but my server, for example from my home wifi, SSL works just fine. It just doesn't work if I try it on the server.
> 
> I have totally no idea what's going on, how can I fix this?

I would guess the iptables rules assume packets go back to
the Ethernet interface (--out-interface eth0) while to
achieve what you're doing it'd need to go to the loopback
interface (lo). In that case, the packets coming back from
httpd to curl would be sent to eth0 and probably lost in
cyberspace...


But then the iptables rules side of transparent proxying has
always been a bit mysterious to me, so maybe this is
wrong...

Y.

**Comment by RadarNyan** [comment]
Turns out this problem has already been solved by #69 

The version I used was v1.17 (1.17-2 from ubuntu package for 16.04) which didn't contain this patch (I misunderstood the date shown on GitHub and thought this patch has been merged before v1.17) Just compiled the latest version (current master branch) and it's working.

**Comment by RadarNyan** [comment]
Wait... I thought that commit solved the problem, but the truth is it breaks transparent proxy completely? Even if I run sslh with ```--transparent```, it works the same as I don't provide ```--trasparent``` argument.

In console I can see (113.229.22.117 is my local ip):

connection from 113.229.22.117:19411 to ```<my_server_ip>```:https forwarded from **172.16.0.1**:44118 to 172.16.0.1:443


If transparent proxy works, the console log looks like this:

connection from 113.229.22.117:16002 to ```<my_server_ip>```:https forwarded from **113.229.22.117**:16002 to 172.16.0.1:443

**Comment by andyboeh** [comment]
Same problem for me, #69 breaks transparent proxy completely

**Comment by andyboeh** [comment]
I just checked in a bit more detail: It's not #69 that breaks transparent proxy, but this commit: https://github.com/yrutschle/sslh/commit/f02ce3821c018719536971dbb1bc1ed1517530a2

(I applied #69 to v1.18 and everything works as expected, it _does_ fix the problem. Then I applied https://github.com/yrutschle/sslh/commit/f02ce3821c018719536971dbb1bc1ed1517530a2 and transparent stopped working)

**Comment by yrutschle** [comment]
Hi Andy,

On Mon, Jan 02, 2017 at 01:31:39PM -0800, andyboeh wrote:
> I just checked in a bit more detail: It's not #69 that breaks transparent proxy, but this commit: https://github.com/yrutschle/sslh/commit/f02ce3821c018719536971dbb1bc1ed1517530a2

Thanks for the investigation -- It appears I accidently
checked in code in a commit that was totally unrelated (I
shall burn in hell eternally).

That commit adds a per-protocol transparent setting. I think
it should work, but I don't remember how much testing I did.
Can you try adding a 'transparent: true' setting to a
specific protocol to check if it's an issue with the global
setting?

Basically you now would have a global setting, and a
per-protocol setting:

transparent: false;  # Global setting
protocols:
(
[...]

# catch anything else TLS -- per-protocol transparent setting
     { name: "tls"; host: "192.168.122.100"; port: "443"; transparent: true; },

[...]
);

Cheers,
Y.

**Comment by andyboeh** [comment]
I'm sorry, I don't want to mess around with setting up a configuration file as I use sslh on a production server without a config file. Maybe @RadarNyan can test that?

**Comment by yrutschle** [comment]
Ok, I reverted f02ce3821c018719536971dbb1bc1ed1517530a2. Sorry about the sloppy commit :/
